go/vt/vtgate/planbuilder/operators/route_planning.go

```diff
@@ -44,73 +44,65 @@ type (
+// This is where a lot of the optimisations of the query plans are done.
+// Here we try to merge query parts into the same route primitives. At the end of this process,
+// all the operators in the tree are guaranteed to be PhysicalOperators
+func TransformToPhysical(ctx *plancontext.PlanningContext, in Operator) (Operator, error) {
+	op, _, err := rewriteBottomUp(ctx, in, func(context *plancontext.PlanningContext, operator Operator) (newOp Operator, changed bool, err error) {
+		switch op := operator.(type) {
+		case *QueryGraph:
+			return optimizeQueryGraph(ctx, op)
+		case *Join:
+			return optimizeJoin(ctx, op)
+		case *Derived:
+			return optimizeDerived(ctx, op)
+		case *SubQuery:
+			return optimizeSubQuery(ctx, op)
+		case *Filter:
+			return optimizeFilter(op)
+		default:
+			return operator, false, nil
 		}
-		sources = append(sources, physOp)
-	}
-	return
-}
+	})
 
-func CreatePhysicalOperator(ctx *plancontext.PlanningContext, opTree Operator) (Operator, error) {
-	switch op := opTree.(type) {
-	case *QueryGraph:
-		return optimizeQueryGraph(ctx, op)
-	case *Join:
-		return optimizeJoin(ctx, op)
-	case *Derived:
-		return optimizeDerived(ctx, op)
-	case *SubQuery:
-		return optimizeSubQuery(ctx, op)
-	case *Filter:
-		return optimizeFilter(ctx, op)
-	case PhysicalOperator:
-		inputs, err := mapToPhys(ctx, op.Inputs())
-		if err != nil {
-			return nil, err
-		}
-		return op.Clone(inputs), nil
-	default:
-		return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "invalid operator tree: %T", op)
+	if err != nil {
+		return nil, err
 	}
-}
 
-func optimizeFilter(ctx *plancontext.PlanningContext, op *Filter) (Operator, error) {
-	src, err := CreatePhysicalOperator(ctx, op.Source)
+	err = VisitTopDown(op, func(op Operator) error {
+		if _, isPhys := op.(PhysicalOperator); !isPhys {
+			return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to transform %T to a physical operator", op)
+		}
+		return nil
+	})
 	if err != nil {
 		return nil, err
 	}
 
-	filter := &Filter{
-		Predicates: op.Predicates,
-	}
+	return op, nil
+}
 
-	if route, ok := src.(*Route); ok {
+func optimizeFilter(op *Filter) (Operator, bool, error) {
+	if route, ok := op.Source.(*Route); ok {
 		// let's push the filter into the route
-		filter.Source = route.Source
-		route.Source = filter
-		return route, nil
+		op.Source = route.Source
+		route.Source = op
+		return route, true, nil
 	}
 
-	filter.Source = src
-
-	return filter, nil
+	return op, false, nil
 }
 
-func optimizeDerived(ctx *plancontext.PlanningContext, op *Derived) (Operator, error) {
-	opInner, err := CreatePhysicalOperator(ctx, op.Source)
-	if err != nil {
-		return nil, err
-	}
+func optimizeDerived(ctx *plancontext.PlanningContext, op *Derived) (Operator, bool, error) {
 
-	innerRoute, ok := opInner.(*Route)
+	innerRoute, ok := op.Source.(*Route)
 	if !ok {
-		return buildDerivedOp(op, opInner), nil
+		return buildDerivedOp(op, op.Source), false, nil
 	}
 
+	// TODO: I THINK THIS SHOULD GO!
```

Comment:
What do you think should go?

Reply:
You are too fast :)

https://github.com/vitessio/vitess/pull/11542/commits/b34a97623d341da6d49d5c3c599caaff6755a775

Reply:
Oh I see now 😛